packages/odata-service-inquirer/src/prompts/datasources/cap-project/cap-helpers.ts

```diff
@@ -0,0 +1,203 @@
+import type { CSN } from '@sap-ux/project-access';
+import {
+    findCapProjects,
+    getCapCustomPaths,
+    getCapModelAndServices,
+    getCdsRoots,
+    readCapServiceMetadataEdmx
+} from '@sap-ux/project-access';
+import { basename, isAbsolute, relative } from 'path';
+import { t } from '../../../i18n';
+import type { CapService, CapServiceChoice } from '../../../types';
+import LoggerHelper from '../../logger-helper';
+import { errorHandler } from '../../prompt-helpers';
+import type { CapProjectChoice, CapProjectPaths, CapProjectRootPath } from './types';
+
+export const enterCapPathChoiceValue = 'enterCapPath';
+
+/**
+ * Search for CAP projects in the specified paths.
+ *
+ * @param paths - The paths used to search for CAP projects
+ * @returns The CAP project paths and the number of folders with the same name
+ */
+async function getCapProjectPaths(
+    paths: string[]
+): Promise<{ capProjectPaths: CapProjectRootPath[]; folderCounts: Map<string, number> }> {
+    const capProjectRoots = await findCapProjects({ wsFolders: paths });
+    const capRootPaths: CapProjectRootPath[] = [];
+    // Keep track of duplicate folder names to append the path to the name when displaying the choices
+    const folderNameCount = new Map<string, number>();
+
+    for (const root of capProjectRoots) {
+        const folderName = basename(root);
+        capRootPaths.push({ folderName, path: root });
+        folderNameCount.set(folderName, (folderNameCount.get(folderName) ?? 0) + 1);
+    }
+    capRootPaths.sort((a, b) => a.folderName.localeCompare(b.folderName));
+    return {
+        capProjectPaths: capRootPaths,
+        folderCounts: folderNameCount
+    };
+}
+
+/**
+ * Search for CAP projects in the specified paths and create prompt choices from the results.
+ * The resulting choices will include an additional entry to enter a custom path.
+ *
+ * @param paths - The paths used to search for CAP projects
+ * @returns The CAP project prompt choices
+ */
+export async function getCapProjectChoices(paths: string[]): Promise<CapProjectChoice[]> {
+    const { capProjectPaths, folderCounts } = await getCapProjectPaths(paths);
+
+    const capChoices: CapProjectChoice[] = [];
+
+    for await (const capProjectPath of capProjectPaths) {
```

Comment:
I would suggest Promise.all() for better performace. The getCapCustomPaths will be called in parallel for each path. Something like this:

```
export async function getCapProjectChoices(paths: string[]): Promise<CapProjectChoice[]> {
    const { capProjectPaths, folderCounts } = await getCapProjectPaths(paths);

    const capChoicesPromises = capProjectPaths.map(async (capProjectPath) => {
        const customCapPaths = await getCapCustomPaths(capProjectPath.path);
        const folderCount = folderCounts.get(capProjectPath.folderName) ?? 1;
        return {
            name: `${capProjectPath.folderName}${folderCount > 1 ? ' (' + capProjectPath.path + ')' : ''}`,
            value: Object.assign(capProjectPath, customCapPaths)
        };
    });

    const capChoices = await Promise.all(capChoicesPromises);

    return [
        ...capChoices,
        {
            name: t('prompts.capProject.enterCapPathChoiceName'),
            value: enterCapPathChoiceValue
        }
    ];
}
```

Reply:
Not sure, would need to do some analysis to see if there's any appreciable difference.